File: opensource/engine-common/src/main/java/com/sos/scheduler/engine/common/xml/XmlUtils.java
package com.sos.scheduler.engine.common.xml;

import com.google.common.collect.ImmutableList;
import com.sos.scheduler.engine.cplusplus.runtime.annotation.ForCpp;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.w3c.dom.Document;
import org.w3c.dom.Element;
import org.w3c.dom.Node;
import org.w3c.dom.NodeList;
import org.xml.sax.InputSource;
import org.xml.sax.SAXException;

import javax.annotation.Nullable;
import javax.xml.parsers.DocumentBuilder;
import javax.xml.parsers.DocumentBuilderFactory;
import javax.xml.parsers.ParserConfigurationException;
import javax.xml.transform.Result;
import javax.xml.transform.Transformer;
import javax.xml.transform.TransformerException;
import javax.xml.transform.TransformerFactory;
import javax.xml.transform.dom.DOMSource;
import javax.xml.transform.stream.StreamResult;
import javax.xml.xpath.XPath;
import javax.xml.xpath.XPathConstants;
import javax.xml.xpath.XPathExpressionException;
import javax.xml.xpath.XPathFactory;
import java.io.*;
import java.nio.charset.Charset;
import java.util.Iterator;

import static com.google.common.base.Objects.firstNonNull;
import static com.google.common.base.Throwables.propagate;
import static javax.xml.transform.OutputKeys.*;
import static org.w3c.dom.Node.DOCUMENT_NODE;

@ForCpp
public final class XmlUtils {
    private static final Logger logger = LoggerFactory.getLogger(XmlUtils.class);
    private static boolean static_xPathNullPointerLogged = false;

    @ForCpp public static Document newDocument() {
        Document result = newDocumentBuilder().newDocument();
        postInitializeDocument(result);
        return result;
    }

    @ForCpp public static Document loadXml(byte[] xml, String encoding) {
        String s;
        try {
            Reader in = new InputStreamReader(new ByteArrayInputStream(xml), Charset.forName(encoding));
            s = com.google.common.io.CharStreams.toString(in);
        } catch (Exception x) { throw propagate(x); }
        return loadXml(new InputStreamReader(new ByteArrayInputStream(xml), Charset.forName(encoding)));
    }

    public static Document loadXml(String xml) {
        return loadXml(new StringReader(xml));
    }

    public static Document loadXml(Reader in) {
        try {
            Document result = newDocumentBuilder().parse(new InputSource(in));
            postInitializeDocument(result);
            return result;
        }
        catch (IOException x) { throw new XmlException(x); }
        catch (SAXException x) { throw new XmlException(x); }
    }

    public static Document loadXml(InputStream in) {
        try {
            Document result = newDocumentBuilder().parse(in);
            postInitializeDocument(result);
            return result;
        }
        catch (IOException x) { throw new XmlException(x); }
        catch (SAXException x) { throw new XmlException(x); }
    }

    private static DocumentBuilder newDocumentBuilder() {
        try {
            DocumentBuilderFactory factory = DocumentBuilderFactory.newInstance();
            factory.setNamespaceAware(true);
            return factory.newDocumentBuilder();
        }
        catch (ParserConfigurationException x) { throw new XmlException(x); }
    }

    private static void postInitializeDocument(Document doc) {
        doc.setXmlStandalone(true);
    }

    @ForCpp public static byte[] toXmlBytes(Node n, String encoding, boolean indent) {
        ByteArrayOutputStream o = new ByteArrayOutputStream();
        writeXmlTo(n, o, Charset.forName(encoding), indent);
        return o.toByteArray();
    }

    public static String toXml(Node n) {
        StringWriter w = new StringWriter();
        writeXmlTo(n, w);
        String result = w.toString();
        return n.getNodeType() == DOCUMENT_NODE? result : removeXmlProlog(result);  // Manche DOM-Implementierung liefert den XML-Prolog.
    }

    private static String removeXmlProlog(String xml) {
        // Prolog muss am Anfang stehen, ohne vorangehende Blanks oder Kommentare
        return xml.startsWith("<?")? xml.replaceFirst("^<[?][xX][mM][lL].+[?][>]\\w*", "") : xml;
    }

    public static void writeXmlTo(Node n, OutputStream o, Charset encoding, boolean indent) {
        writeXmlTo(n, new StreamResult(o), encoding, indent);
    }

    public static void writeXmlTo(Node n, Writer w) {
        writeXmlTo(n, new StreamResult(w), null, false);
    }

    private static void writeXmlTo(Node n, Result r, @Nullable Charset encoding, boolean indent) {
        try {
            Transformer transformer = TransformerFactory.newInstance().newTransformer();
            if (encoding != null)
                transformer.setOutputProperty(ENCODING, encoding.name());
            transformer.setOutputProperty(OMIT_XML_DECLARATION, n.getNodeType() == DOCUMENT_NODE? "no" : "yes");   //TODO Funktioniert nur mit org.jdom?
            if (indent) {
                transformer.setOutputProperty(INDENT, "yes");
                transformer.setOutputProperty("{http://xml.apache.org/xslt}indent-amount", "4");
            }
            transformer.transform(new DOMSource(n), r);
        } catch (TransformerException x) { throw new XmlException(x); }
    }

    @ForCpp public static boolean booleanXmlAttribute(Element xmlElement, String attributeName, boolean defaultValue) {
        String value = xmlElement.getAttribute(attributeName);
        Boolean result = booleanOrNullOf(value, defaultValue);
        if (result == null)
            throw new RuntimeException("Invalid Boolean value in <"+ xmlElement.getNodeName() +" "+ attributeName + "=" + xmlQuoted(value) + ">");
        return result;
    }

    @Nullable private static Boolean booleanOrNullOf(String s, boolean deflt) {
        return s.equals("true")? true :
               s.equals("false")? false :
               s.equals("1")? true :
               s.equals("0")? false :
               s.isEmpty()? deflt : null;
    }

    @ForCpp public static int intXmlAttribute(Element xmlElement, String attributeName) {
        return intXmlAttribute(xmlElement, attributeName, (Integer)null);
    }

    @ForCpp public static int intXmlAttribute(Element xmlElement, String attributeName, @Nullable Integer defaultValue) {
        String value = xmlAttribute(xmlElement, attributeName, "");
        if (!value.isEmpty()) {
            try {
                return Integer.parseInt(value);
            } catch (NumberFormatException x) {
                throw new RuntimeException("Invalid numeric value in <"+ xmlElement.getNodeName() +" "+ attributeName +"="+ xmlQuoted(value) +">", x);
            }
        } else {
            if (defaultValue == null)
                throw missingAttributeException(xmlElement, attributeName);
            return defaultValue;
        }
    }

    @ForCpp public static String xmlAttribute(Element xmlElement, String attributeName, @Nullable String defaultValue) {
        String result = xmlElement.getAttribute(attributeName);
        if (!result.isEmpty())
            return result;
        else {
            if (defaultValue == null)
                throw missingAttributeException(xmlElement, attributeName);
            return defaultValue;
        }
    }

    private static RuntimeException missingAttributeException(Element e, String attributeName) {
        return new RuntimeException("Missing attribute <"+ e.getNodeName() +" "+ attributeName +"=...>");
    }

    public static Element elementXPath(Node baseNode, String xpathExpression) {
        Element result = elementXPathOrNull(baseNode, xpathExpression);
        if (result == null)  throw new XmlException("XPath does not return an element: "+ xpathExpression);
        return result;
    }

    @Nullable public static Element elementXPathOrNull(Node baseNode, String xpathExpression) {
        try {
            return (Element)newXPath().evaluate(xpathExpression, baseNode, XPathConstants.NODE);
        } catch (XPathExpressionException x) { throw new XmlException(x); }
    }

    public static ImmutableList<Element> elementsXPath(Node baseNode, String xpathExpression) {
        try {
            return elementListFromNodeList((NodeList)newXPath().evaluate(xpathExpression, baseNode, XPathConstants.NODESET));
        } catch (XPathExpressionException x) { throw new XmlException(x); }
    }

    public static ImmutableList<Element> elementListFromNodeList(NodeList list) {
        ImmutableList.Builder<Element> result = ImmutableList.builder();
        for (int i = 0; i < list.getLength(); i++)  result.add((Element)list.item(i));
        return result.build();
    }

    public static String stringXPath(Node baseNode, String xpathExpression) {
        try {
            String result = (String)newXPath().evaluate(xpathExpression, baseNode, XPathConstants.STRING);
            if (result == null)  throw new XmlException("XPath does not match: "+ xpathExpression);
            return result;
        } catch (XPathExpressionException x) { throw new XmlException(x); }
    }

    public static String stringXPath(Node baseNode, String xpathExpression, String deflt) {
        try {
            String result = (String)newXPath().evaluate(xpathExpression, baseNode, XPathConstants.STRING);
            return firstNonNull(result, deflt);
        } catch (XPathExpressionException x) { throw new XmlException(x); }
    }

    public static boolean booleanXPath(Node baseNode, String xpathExpression) {
        try {
            return (Boolean)newXPath().evaluate(xpathExpression, baseNode, XPathConstants.BOOLEAN);
        } catch (XPathExpressionException x) { throw new XmlException(x); }
    }

    public static ImmutableList<Element> childElements(Element element) {
        return ImmutableList.copyOf(new SiblingElementIterator(element.getFirstChild()));
    }

    @Nullable public static Element childElementOrNull(Element e, String name) {
        Iterator<Element> i = new NamedChildElements(name, e).iterator();
        return i.hasNext()? i.next() : null;
    }

    @ForCpp public static NodeList xpathNodeList(Node baseNode, String xpathExpression) {
        try {
            XPath xpath = newXPath();
            return (NodeList)xpath.evaluate( xpathExpression, baseNode, XPathConstants.NODESET );
        } catch (XPathExpressionException x) { throw new XmlException( x ); }   // Programmfehler
    }

    @ForCpp public static Node xpathNode(Node baseNode, String xpathExpression) {
        try {
            XPath xpath = newXPath();
            return (Node)xpath.evaluate(xpathExpression, baseNode, XPathConstants.NODE);
        } catch (XPathExpressionException x) { throw new XmlException( x ); }   // Programmfehler
    }

    public static XPath newXPath() {
        return newXPathFactory().newXPath();
    }

    private static XPathFactory newXPathFactory() {
        try {
            return XPathFactory.newInstance();
        } catch (NullPointerException e) {
            return workaroundNewXPathFactory(e);
        }
    }

    private static XPathFactory workaroundNewXPathFactory(NullPointerException e) {
        // JSSIXFOUR-8: NullPointerException in javax.xml.xpath.XPathFactoryFinder, wenn Scheduler als Dienst läuft
        String workAroundClassName = "com.sun.org.apache.xpath.internal.jaxp.XPathFactoryImpl";
        if (!static_xPathNullPointerLogged) {
            logger.debug("Trying to use "+workAroundClassName+" as a workaround after {}", e, e);
            static_xPathNullPointerLogged = true;
        }
        try {
            @SuppressWarnings("unchecked")
            XPathFactory result = ((Class<XPathFactory>)Class.forName(workAroundClassName)).newInstance();
            return result;
        } catch (Throwable ee) {
            logger.debug("Workaround failed", ee);
            // Ab Java 7: e.addSuppressed(ee);
            throw e;
        }
    }

    public static String xmlQuoted(String value) {
        StringBuilder result = new StringBuilder(value.length() + 20);

        for (int i = 0; i < value.length(); i++) {
            char c = value.charAt( i );
            switch (c) {
                case '"': result.append("&quot;");  break;
                case '&': result.append("&amp;" );  break;
                case '<': result.append("&lt;"  );  break;
                default:  result.append(c);
            }
        }

        return "\"" + result + "\"";
    }

    private static final class XmlException extends RuntimeException {
        private XmlException(Exception x) {
            super(x);
        }

        private XmlException(String s) {
            super(s);
        }
    }

    private XmlUtils() {}
}

File: opensource/engine-common/src/main/scala/com/sos/scheduler/engine/common/xml/XmlUtils.scala
package com.sos.scheduler.engine.common.xml

import com.google.common.base.Objects.firstNonNull
import com.google.common.collect.ImmutableList
import com.sos.scheduler.engine.common.scalautil.ModifiedBy.modifiedBy
import com.sos.scheduler.engine.common.scalautil.StringWriters.writingString
import com.sos.scheduler.engine.common.scalautil.{ModifiedBy, Logger}
import com.sos.scheduler.engine.common.scalautil.ScalaThreadLocal._
import com.sos.scheduler.engine.cplusplus.runtime.annotation.ForCpp
import java.io._
import java.nio.charset.Charset
import javax.annotation.Nullable
import javax.xml.parsers.DocumentBuilderFactory
import javax.xml.transform.OutputKeys._
import javax.xml.transform.dom.DOMSource
import javax.xml.transform.stream.StreamResult
import javax.xml.transform.{Result, TransformerFactory}
import javax.xml.xpath.{XPathConstants, XPathFactory}
import org.w3c.dom.Node.DOCUMENT_NODE
import org.w3c.dom.{Document, Element, Node, NodeList}
import org.xml.sax.InputSource
import scala.sys.error

@ForCpp object XmlUtils {
  private val logger = Logger(getClass)
  private lazy val xPathFactory = newXPathFactory()
  private lazy val xPath = threadLocal { xPathFactory.newXPath() }
  private val documentBuilder = threadLocal { (DocumentBuilderFactory.newInstance() modifiedBy { _.setNamespaceAware(true) }).newDocumentBuilder() }
  private val transformerFactory = threadLocal { TransformerFactory.newInstance() }
  private var static_xPathNullPointerLogged = false

  @ForCpp def newDocument(): Document = {
    val result = documentBuilder.newDocument()
    postInitializeDocument(result)
    result
  }

  def prettyXml(xml: String): String =
    writingString { w => writeXmlTo(loadXml(xml), new StreamResult(w), None, indent=true) }

  @ForCpp def loadXml(xml: Array[Byte], encoding: String): Document =
    loadXml(new InputStreamReader(new ByteArrayInputStream(xml), Charset.forName(encoding)))

  def loadXml(xml: String): Document =
    loadXml(new StringReader(xml))

  def loadXml(in: Reader): Document =
    documentBuilder.parse(new InputSource(in)) modifiedBy postInitializeDocument

  def loadXml(in: InputStream): Document = {
    val result = documentBuilder.parse(in)
    postInitializeDocument(result)
    result
  }

  private def postInitializeDocument(doc: Document) {
    doc.setXmlStandalone(true)
  }

  @ForCpp def toXmlBytes(n: Node, encoding: String, indent: Boolean): Array[Byte] = {
    val o = new ByteArrayOutputStream
    writeXmlTo(n, o, Charset.forName(encoding), indent)
    o.toByteArray
  }

  def toXml(n: Node): String = {
    val result = writingString { w => writeXmlTo(n, w) }
    if (n.getNodeType == DOCUMENT_NODE) result
    else removeXmlProlog(result)
  }

  private def removeXmlProlog(xml: String) =
    if (xml startsWith "<?") xml.replaceFirst("^<[?][xX][mM][lL].+[?][>]\\w*", "") else xml

  def writeXmlTo(n: Node, o: OutputStream, encoding: Charset, indent: Boolean) {
    writeXmlTo(n, new StreamResult(o), Some(encoding), indent)
  }

  def writeXmlTo(n: Node, w: Writer) {
    writeXmlTo(n, new StreamResult(w), None)
  }

  private def writeXmlTo(n: Node, r: Result, encoding: Option[Charset], indent: Boolean = false) {
    val transformer = transformerFactory.newTransformer()
    for (o <- encoding) transformer.setOutputProperty(ENCODING, o.name)
    transformer.setOutputProperty(OMIT_XML_DECLARATION, if (n.getNodeType == DOCUMENT_NODE) "no" else "yes")
    if (indent) {
      transformer.setOutputProperty(INDENT, "yes")
      transformer.setOutputProperty("{http://xml.apache.org/xslt}indent-amount", "4")
    }
    transformer.transform(new DOMSource(n), r)
  }

  @ForCpp def booleanXmlAttribute(xmlElement: Element, attributeName: String, defaultValue: Boolean): Boolean = {
    val value = xmlElement.getAttribute(attributeName)
    booleanOptionOf(value, defaultValue) getOrElse error(s"Invalid Boolean value in <${xmlElement.getNodeName} $attributeName=${xmlQuoted(value)}>")
  }

  private def booleanOptionOf(s: String, deflt: Boolean): Option[Boolean] = s match {
    case "true" => Some(true)
    case "false" => Some(false)
    case "1" => Some(true)
    case "0" => Some(false)
    case _ if (s.isEmpty) => Some(deflt)
    case _ => None
  }

  @ForCpp def intXmlAttribute(xmlElement: Element, attributeName: String): Int =
    intXmlAttribute(xmlElement, attributeName, null.asInstanceOf[Integer])

  @ForCpp def intXmlAttribute(xmlElement: Element, attributeName: String, @Nullable defaultValue: Integer): Int = {
    val value = xmlAttribute(xmlElement, attributeName, "")
    if (!value.isEmpty) {
      try Integer.parseInt(value)
      catch {
        case x: NumberFormatException =>
          throw new RuntimeException(s"Invalid numeric value in <${xmlElement.getNodeName} $attributeName=${xmlQuoted(value)}>", x)
      }
    }
    else {
      if (defaultValue == null) throw missingAttributeException(xmlElement, attributeName)
      defaultValue
    }
  }

  @ForCpp def xmlAttribute(xmlElement: Element, attributeName: String, @Nullable defaultValue: String): String = {
    val result = xmlElement.getAttribute(attributeName)
    if (!result.isEmpty) result
    else {
      if (defaultValue == null) throw missingAttributeException(xmlElement, attributeName)
      defaultValue
    }
  }

  private def missingAttributeException(e: Element, attributeName: String) =
    new RuntimeException(s"Missing attribute <${e.getNodeName} $attributeName=...>")

  def elementXPath(baseNode: Node, xpathExpression: String): Element =
    elementXPathOption(baseNode, xpathExpression) getOrElse error(s"XPath does not return an element: $xpathExpression")

  @Nullable def elementXPathOrNull(baseNode: Node, xpathExpression: String): Element =
    elementXPathOption(baseNode, xpathExpression).orNull

  def elementXPathOption(baseNode: Node, xpathExpression: String): Option[Element] =
    Option(xPath.evaluate(xpathExpression, baseNode, XPathConstants.NODE).asInstanceOf[Element])

  def elementsXPath(baseNode: Node, xpathExpression: String) =
    elementListFromNodeList(xPath.evaluate(xpathExpression, baseNode, XPathConstants.NODESET).asInstanceOf[NodeList])

  def elementListFromNodeList(list: NodeList): IndexedSeq[Element] =
    0 until list.getLength map list.item map { _.asInstanceOf[Element] }

  def stringXPath(baseNode: Node, xpathExpression: String): String = {
      val result = xPath.evaluate(xpathExpression, baseNode, XPathConstants.STRING).asInstanceOf[String]
      if (result == null) error("XPath does not match: " + xpathExpression)
      result
  }

  def stringXPath(baseNode: Node, xpathExpression: String, deflt: String): String = {
    val result = xPath.evaluate(xpathExpression, baseNode, XPathConstants.STRING).asInstanceOf[String]
    firstNonNull(result, deflt)
  }

  def booleanXPath(baseNode: Node, xpathExpression: String): Boolean =
    xPath.evaluate(xpathExpression, baseNode, XPathConstants.BOOLEAN).asInstanceOf[Boolean]

  def childElements(element: Element): ImmutableList[Element] =
    ImmutableList.copyOf(new SiblingElementIterator(element.getFirstChild))

  @Nullable def childElementOrNull(e: Element, name: String): Element = {
    val i = new NamedChildElements(name, e).iterator
    if (i.hasNext) i.next else null
  }

  @ForCpp def xpathNodeList(baseNode: Node, xpathExpression: String): NodeList = {
    xPath.evaluate(xpathExpression, baseNode, XPathConstants.NODESET).asInstanceOf[NodeList]
  }

  @ForCpp def xpathNode(baseNode: Node, xpathExpression: String): Node =
    xPath.evaluate(xpathExpression, baseNode, XPathConstants.NODE).asInstanceOf[Node]

  private def newXPathFactory(): XPathFactory = {
    try XPathFactory.newInstance
    catch { case e: NullPointerException => workaroundNewXPathFactory(e) }
  }

  private def workaroundNewXPathFactory(e: NullPointerException): XPathFactory = {
    val workAroundClassName = "com.sun.org.apache.xpath.internal.jaxp.XPathFactoryImpl"
    if (!static_xPathNullPointerLogged) {
      logger.debug(s"Trying to use $workAroundClassName as a workaround after $e", e)
      static_xPathNullPointerLogged = true
    }
    try {
      @SuppressWarnings(Array("unchecked")) val result = (Class.forName(workAroundClassName).asInstanceOf[Class[XPathFactory]]).newInstance
      result
    }
    catch {
      case ee: Throwable => {
        logger.debug("Workaround failed", ee)
        throw e
      }
    }
  }

  def xmlQuoted(value: String): String = {
    val result = new StringBuilder(value.length + 20)
    value foreach {
      _ match {
        case '"' => result.append("&quot;")
        case '&' => result.append("&amp;")
        case '<' => result.append("&lt;")
        case o => result.append(o)
      }
    }
    "\""+ result +"\""
  }
}
